### PR TITLE
[Bug][Structured Outputs] Fix bug that leads to unconstrained generations with structural tags

### PR DESCRIPTION
**Pull request by rishitdholakia13**
With structural tags, a reasoning parser, and speculative decoding, we sometimes get logits that are not grammar-masked (allow-all bitmask), which leads to unconstrained token generation when we expect guided decoding.

What goes wrong

- [reasoning_ended](https://github.com/vllm-project/vllm/blob/379f0ec369d300d6fb45983882f1a4d548065d73/vllm/v1/structured_output/__init__.py#L355) (which gates both should_fill_bitmask and [should_advance](https://github.com/vllm-project/vllm/blob/379f0ec369d300d6fb45983882f1a4d548065d73/vllm/v1/structured_output/__init__.py#L321)) is updated only after a decode step finishes—when should_advance runs during update_from_output.

- While should_advance is false (still treating the stream as “reasoning”), the scheduler does not call grammar.accept_tokens(...) on the new token IDs. The grammar matcher therefore does not advance through prefix / trigger / begin states for those tokens.

- Speculative decoding can commit many token IDs in one step. Those IDs can include trigger (and following) text which are necessary for the FSM advancement. 

- On the next step, the matcher is still at an earlier state while the trigger tags are now a part of the input ids. The next accept_tokens calls then see tokens that do not match the trigger tags (since the accept_tokens are called on the accepted output tokens) , so structured decoding is not enabled in the following calls as well.

### CODE REVIEW

File: vllm/v1/structured_output/__init__.py [hunk 1]
```diff
@@ -354,6 +355,17 @@ def should_advance(self, request: "Request") -> bool:
             # next pass
             structured_req.reasoning_ended = True
 
+            # Only treat reasoning_ended as allowing FSM advance for speculative
+            # decoding with structural-tag constraints; other paths keep prior
+            # streaming gating without this shortcut.
+            if (
+                structured_req.reasoning_ended
```
[the remark this answers is not in the document]
Comment (cjackal) [reply vllm/v1/structured_output/__init__.py]: Gemini's review here seems valid, we already set `structured_req.reasoning_end = True` at L356.

[the remark this answers is not in the document]
Comment (rishitdholakia13) [reply vllm/v1/structured_output/__init__.py]: Will make the change

**Comment by rishitdholakia13** [comment]
@aarnphm Could you have a look at the fix. Thank you !

File: vllm/v1/structured_output/__init__.py [hunk 2]
```diff
@@ -354,6 +355,16 @@ def should_advance(self, request: "Request") -> bool:
...
 
+            # Only treat reasoning_ended as allowing FSM advance for speculative
+            # decoding with structural-tag constraints; other paths keep prior
+            # streaming gating without this shortcut.
+            if (
+                self.vllm_config.speculative_config is not None
+                and structured_req.structured_output_key[0]
+                == StructuredOutputOptions.STRUCTURAL_TAG
+            ):
+                return True
```
Comment (mgoin) [inline vllm/v1/structured_output/__init__.py]: I think we can drop the redundant reasoning_ended check just set above now, and please note in the comment why structural-tag is the only safe case here

**Review by mgoin — COMMENTED** [review]
Thanks for this fix! I believe your analysis is right and fix works for the targeted case. Could you please add a regression test? Probably just mirror test_should_advance_reasoning_just_ended with speculative_config set + STRUCTURAL_TAG key.
Also worth filing a follow-up issue for JSON/regex + spec + reasoning

File: vllm/v1/structured_output/__init__.py (hunk 2 above)
Comment (rishitdholakia13) [reply vllm/v1/structured_output/__init__.py]: Thankd for the comments ! If you mean by removing the reddunded reasoning_ended check [here](https://github.com/vllm-project/vllm/pull/42452/changes#diff-a0726d18a97a553ef12f020d2ab9d3878129690b5d379ad5d1de45e4d9a5af27L341-L342). We should probably keep it, as it returns True for FSM advancement once reasoning ends. Since the change made in this PR is just the one particular step transition. Just keeping it there would Fail advancement of FSM in the following steps.

**Review by mgoin — APPROVED** [review]
